File: 백준/Silver/9012. 괄호/괄호.py
import sys
re = int(sys.stdin.readline())
for _ in range(re):
    sentence = str(sys.stdin.readline().strip())
    stack = []
    for i in sentence:
        if i == '(':
            stack.append('(')
        elif i == ')':
            if len(stack) != 0 and stack[-1] == '(':
                stack.pop()
            else:
                stack.append(i)
    if len(stack) == 0:
        print('YES')
    else:
        print('NO')

File: 백준/Silver/9012. 괄호/괄호.cc
#include <iostream>
#include <string>
#include <stack>
using namespace std;

int main(){
	stack<char> st;
	int size;
	cin >> size;

	for (int i = 0; i < size; i++){
		string s;
		cin >> s;
		for (int j = 0; j < s.length(); j++){
			if (s[j] == '(' || st.empty()) st.push(s[j]);
			else if(st.top() == '(')st.pop();
		}

		if (st.empty()) cout << "YES" << endl;
		else cout << "NO" << endl;
		while (!st.empty()) st.pop();
	}
}
